File: site/iot/sensors/oscillator.js
// generator of correct traces for oscillator.rml: anomaly detection for a damped harmonic oscillator with a distance sensor 

'use strict';

function getOscillation(t){
    console.log(amplitude*(Math.exp(-zeta*omega0*t)*Math.sin(omega1*t+phase)));
    return amplitude*(Math.exp(-zeta*omega0*t)*Math.sin(omega1*t+phase));
}

function sensor(x) {
    event_counter++;
    var res = {position:getOscillation(x), time : x};
    return res;
}

var trace_size = Number(process.argv[2]) || 100;
var event_counter = 0;
var amplitude = 0.5;
var m=1;
var phase=0;
var k=5000;
var c=18.6*Math.pow(10,-6);
var e=Math.exp(1);
var zeta=c/(2*Math.sqrt(m*k));
var omega0=Math.sqrt(k/m);
var omega1=omega0*Math.sqrt(1-zeta*zeta);

// console.log(`e:${e}`);
// console.log(`c:${c}`);
// console.log(`zeta:${zeta}`);
// console.log(`omega0:${omega0}`);
// console.log(`omega1:${omega1}`);
while(event_counter<=trace_size){
    sensor(event_counter);
}

File: site/iot/sensors/oscillator.pl
:- module(spec, [(trace_expression/2), (match/2)]).
:- use_module(monitor(deep_subdict)).
:- use_module(library(clpr)).
match(_event, sensor_et(Pos, Time)) :- deep_subdict(_event, _{event:"func_post", name:"sensor", res:_{position:Pos, time:Time}}).
match(_event, check_dots_et(Pos1, Time1, Pos2, Time2)) :- match(_event, sensor_et(Pos2, Time2)),
	{(((((((((((((((E =:= 2.718),(K =:= 5000)),(C =:= (18.6 * (10 ^ (-6))))),(M =:= 1)),(Phase =:= 0)),(Error =:= (10 ^ (-5)))),(Zeta =:= (C / (2 * ((M * K) ^ 0.5))))),(Omega0 =:= ((K / M) ^ 0.5))),(Omega1 =:= (Omega0 * ((1 - (Zeta ^ 2)) ^ 0.5)))),(Delta1 =:= (Pos1 - (Amplitude * ((E ^ (((-Zeta) * Omega0) * Time1)) * sin(((Omega1 * Time1) + Phase))))))),(Delta2 =:= (Pos2 - (Amplitude * ((E ^ (((-Zeta) * Omega0) * Time2)) * sin(((Omega1 * Time2) + Phase))))))),(Delta1 =< Error)),(Delta1 >= (-Error))),(Delta2 =< Error)),(Delta2 >= (-Error)))}.
match(_event, relevant_et) :- match(_event, sensor_et(_, _)).
match(_event, any_et) :- deep_subdict(_event, _{}).
match(_event, none_et) :- not(match(_event, any_et)).
trace_expression('Main', Main) :- (Main=((relevant_et>>clos(var(pos1, var(time1, (sensor_et(var(pos1), var(time1))*app(CheckDot, [var(pos1), var(time1)]))))));1)),
	(CheckDot=gen([pos1, time1], var(pos2, var(time2, (check_dots_et(var(pos1), var(time1), var(pos2), var(time2))*app(CheckDot, [var(pos2), var(time2)])))))).
